firewalld: Implement zone operations

kaechele opened the pull request:
##### SUMMARY
Zones are removed or added when no other operations are used in
conjunction with the keywords 'present' or 'absent'.

This leads to a logical and natural syntax of:

- firewalld:
    zone: foo
    state: present

for adding or removing zones.

Signed-off-by: Felix Kaechele <felix@kaechele.ca>

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
firewalld module

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.4.1.0
  config file = /home/felix/.ansible.cfg
  configured module search path = [u'/home/felix/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.14 (default, Nov  3 2017, 10:55:25) [GCC 7.2.1 20170915 (Red Hat 7.2.1-2)]
```

kaechele committed:
firewalld: Implement zone operations

Zones are removed or added when no other operations are used in
conjunction with the keywords 'present' or 'absent'.

This leads to a logical and natural syntax of:

- firewalld:
    zone: foo
    state: present

for adding or removing zones.

Signed-off-by: Felix Kaechele <felix@kaechele.ca>

maxamillion commented on lib/ansible/modules/system/firewalld.py:
```diff
@@ -926,6 +977,27 @@ def main():
+        )
+
+        transaction.enabled_values = ['present']
+        transaction.disabled_values = ['absent']
```
What's the motivation for setting this after instantiation instead of passing these as extra named params to `__init__`?

maxamillion commented on lib/ansible/modules/system/firewalld.py:
```diff
@@ -926,6 +977,27 @@ def main():
+            permanent=True,
+            immediate=False,
+            fw_offline=fw_offline
+        )
```
I'd like to see the `NotImplementedError` exceptions be handled here so the user gets a clean error message via `module.fail_json` in the event one of those occurs.

kaechele commented:
I've updated my repo. When all is fine I'll squash the commits.

maxamillion commented:
@kaechele please make the bot happy :)

kaechele committed:
firewalld: zone ops: addressed review concerns

- Added more documentation on the peculiarities of the zone operations
- Output meaningful error messages when trying to use zones incorrectly

Signed-off-by: Felix Kaechele <felix@kaechele.ca>

maxamillion commented:
shipit
